staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit_test.go

```diff
@@ -984,3 +704,71 @@ func TestAuditLevelNone(t *testing.T) {
 		t.Errorf("Generated events, but should not have: %#v", sink.events)
 	}
 }
+
+func TestAuditIDHttpHeader(t *testing.T) {
+	for _, test := range []struct {
+		desc           string
+		requestHeader  string
+		level          auditinternal.Level
+		expectedHeader bool
+	}{
+		{
+			"no http header when there is no audit",
+			"",
+			auditinternal.LevelNone,
+			false,
+		},
+		{
+			"no http header when there is no audit even the request header specified",
+			uuid.NewRandom().String(),
+			auditinternal.LevelNone,
+			false,
+		},
+		{
+			"server generated header",
+			"",
+			auditinternal.LevelRequestResponse,
+			true,
+		},
+		{
+			"user provided header",
+			uuid.NewRandom().String(),
+			auditinternal.LevelRequestResponse,
+			true,
+		},
+	} {
+		sink := &fakeAuditSink{}
+		var handler http.Handler
+		handler = http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+			w.WriteHeader(200)
+		})
+		policyChecker := policy.FakeChecker(test.level)
+		handler = WithAudit(handler, &fakeRequestContextMapper{
+			user: &user.DefaultInfo{Name: "admin"},
+		}, sink, policyChecker, nil)
+
+		req, _ := http.NewRequest("GET", "/api/v1/namespaces/default/pods", nil)
+		req.RemoteAddr = "127.0.0.1"
+		if test.requestHeader != "" {
+			req.Header.Add("Audit-ID", test.requestHeader)
+		}
+
+		w := httptest.NewRecorder()
+		handler.ServeHTTP(w, req)
+		resp := w.Result()
+		if test.expectedHeader {
+			if resp.Header.Get("Audit-ID") == "" {
+				t.Errorf("[%s] expected Audit-ID http header returned, but not returned", test.desc)
```

Comment:
Instead of putting [%s] in front of each string, you can use the following approach:

```
for _, tc := range tcs {
  t.Run(tc.desc, func() {
    ...
  })
}
```

Then you can use `t.Failf` to fail the test without continuation